mm2src/coins/qrc20.rs

```diff
@@ -0,0 +1,1004 @@
+            })?;
+        let prev_script = ScriptBuilder::build_p2pkh(&self.utxo.my_address.hash);
+        let signed = try_s!(sign_tx(
+            unsigned,
+            &self.utxo.key_pair,
+            prev_script,
+            self.utxo.signature_version,
+            self.utxo.fork_id
+        ));
+        Ok(GenerateQrc20TxResult {
+            signed,
+            miner_fee: data.fee_amount,
+            gas_fee,
+        })
+    }
+
+    fn transfer_output(
+        &self,
+        to_addr: H160,
+        amount: U256,
+        gas_limit: u64,
+        gas_price: u64,
+    ) -> Result<ContractCallOutput, String> {
+        let function = try_s!(ERC20_CONTRACT.function("transfer"));
+        let params = try_s!(function.encode_input(&[Token::Address(to_addr), Token::Uint(amount)]));
+
+        let script_pubkey = try_s!(generate_contract_call_script_pubkey(
+            &params,
+            gas_limit,
+            gas_price,
+            &self.contract_address,
+        ))
+        .to_bytes();
+
+        Ok(ContractCallOutput {
+            value: OUTPUT_QTUM_AMOUNT,
+            script_pubkey,
+            gas_limit,
+            gas_price,
+        })
+    }
+}
+
+#[cfg_attr(test, mockable)]
+#[async_trait]
+impl UtxoCommonOps for Qrc20Coin {
+    async fn get_tx_fee(&self) -> Result<ActualTxFee, JsonRpcError> { utxo_common::get_tx_fee(&self.utxo).await }
+
+    async fn get_htlc_spend_fee(&self) -> Result<u64, String> { utxo_common::get_htlc_spend_fee(self).await }
+
+    fn addresses_from_script(&self, script: &Script) -> Result<Vec<UtxoAddress>, String> {
+        utxo_common::addresses_from_script(&self.utxo, script)
+    }
+
+    fn denominate_satoshis(&self, satoshi: i64) -> f64 { utxo_common::denominate_satoshis(&self.utxo, satoshi) }
+
+    fn my_public_key(&self) -> &Public { self.utxo.key_pair.public() }
+
+    fn display_address(&self, address: &UtxoAddress) -> Result<String, String> {
+        utxo_common::display_address(&self.utxo, address)
+    }
+
+    fn try_address_from_str(&self, from: &str) -> Result<UtxoAddress, String> {
+        let utxo_err = match UtxoAddress::from_str(from) {
+            Ok(addr) => {
+                let is_p2pkh =
+                    addr.prefix == self.utxo.pub_addr_prefix && addr.t_addr_prefix == self.utxo.pub_t_addr_prefix;
+                if is_p2pkh {
+                    return Ok(addr);
+                }
+                "Address has invalid prefixes".to_string()
+            },
+            Err(e) => e.to_string(),
+        };
+        let qrc20_err = match qrc20_addr_from_str(from) {
+            Ok(qrc20_addr) => return Ok(self.utxo_address_from_qrc20(qrc20_addr)),
+            Err(e) => e,
+        };
+        ERR!(
+            "error on parse wallet address: {:?}, error on parse contract address: {:?}",
+            utxo_err,
+            qrc20_err,
+        )
+    }
+
+    fn address_from_str(&self, address: &str) -> Result<UtxoAddress, String> {
```

Comment:
The naming is confusing a bit here. E.g. there are `From` trait that has `fn from()` that returns `Self` and `TryFrom` trait with fn `try_from()` that returns `Result<Self, E>`. But we have `address_from_str` and `try_address_from_str` that both return `Result<UtxoAddress, String>`. Please consider renaming or making 1 fn if possible.

Reply:
Thank you for your comment. Please check the last commit :)